examples/D3D12Binding.cpp

```diff
@@ -103,65 +102,147 @@ class D3D12Binding : public BackendBinding {
+                    nullptr,
+                    IID_PPV_ARGS(&commandList)
+                ));
+
+                // Transition the first frame to be a render target
+
+                D3D12_RESOURCE_BARRIER resourceBarrier;
+                resourceBarrier.Transition.pResource = renderTargetResources[renderTargetIndex].Get();
+                resourceBarrier.Transition.StateBefore = D3D12_RESOURCE_STATE_PRESENT;
+                resourceBarrier.Transition.StateAfter = D3D12_RESOURCE_STATE_RENDER_TARGET;
+                resourceBarrier.Transition.Subresource = D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES;
+                resourceBarrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
+                resourceBarrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
+                commandList->ResourceBarrier(1, &resourceBarrier);
+                ASSERT_SUCCESS(commandList->Close());
+                ID3D12CommandList* commandLists[] = { commandList.Get() };
+                commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            }
+
+            D3D12_CPU_DESCRIPTOR_HANDLE renderTargetViewHandle = renderTargetViewHeap->GetCPUDescriptorHandleForHeapStart();
+            renderTargetViewHandle.ptr += rtvDescriptorSize * renderTargetIndex;
+            backend::d3d12::SetNextRenderTargetDescriptor(backendDevice, renderTargetViewHandle);
         }
+
         void SwapBuffers() override {
-            // Present the rendered frame
+            // Transition current frame's render target for presenting
+            {
+                ASSERT_SUCCESS(commandList->Reset(commandAllocators[renderTargetIndex].Get(), nullptr));
+                D3D12_RESOURCE_BARRIER resourceBarrier;
+                resourceBarrier.Transition.pResource = renderTargetResources[renderTargetIndex].Get();
+                resourceBarrier.Transition.StateBefore = D3D12_RESOURCE_STATE_RENDER_TARGET;
+                resourceBarrier.Transition.StateAfter = D3D12_RESOURCE_STATE_PRESENT;
+                resourceBarrier.Transition.Subresource = D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES;
+                resourceBarrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
+                resourceBarrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
+                commandList->ResourceBarrier(1, &resourceBarrier);
+                ASSERT_SUCCESS(commandList->Close());
+                ID3D12CommandList* commandLists[] = { commandList.Get() };
+                commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            }
+
             ASSERT_SUCCESS(swapChain->Present(1, 0));
+            ASSERT_SUCCESS(commandQueue->Signal(fence.Get(), currentFrameNumber));
 
-            const uint64_t currentFenceValue = fenceValues[frameIndex];
-            ASSERT_SUCCESS(commandQueue->Signal(fence.Get(), currentFenceValue));
+            // Transition last frame's render target back to being a render target
```

Comment:
This should be done before the signal, otherwise when the fence value is updated, the commands in commandAllocators[renderTargetIndex] might still be in flight.